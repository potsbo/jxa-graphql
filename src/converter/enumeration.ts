import { EnumTypeDefinitionNode, EnumValueDefinitionNode, Kind, StringValueNode } from "graphql";
import camelCase from "camelcase";
import { EnumDefinition, Environment } from "./sdef";

export class EnumBuilder {
  private e: EnumDefinition;
  constructor(e: EnumDefinition) {
    this.e = e;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  build = (_: Environment): EnumTypeDefinitionNode[] => {
    const values: EnumValueDefinitionNode[] = this.e.enumerator.map((e) => {
      const desc: StringValueNode | undefined = e.$.description
        ? {
            kind: Kind.STRING,
            value: e.$.description,
            block: true,
          }
        : undefined;
      return {
        kind: Kind.ENUM_VALUE_DEFINITION,
        description: desc,
        name: {
          kind: Kind.NAME,
          value: e.$.name.replace(/ /g, "_").replace(/&/g, "_AND_").toUpperCase(),
        },
      };
    });
    return [
      {
        kind: Kind.ENUM_TYPE_DEFINITION,
        name: {
          kind: Kind.NAME,
          value: camelCase(this.e.$.name, { pascalCase: true }),
        },
        values,
      },
    ];
  };
}
