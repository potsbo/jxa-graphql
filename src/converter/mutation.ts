import { InputValueDefinitionNode, Kind } from "graphql";
import { ContentDefinition, ElementDefinition, PropertyDefinition } from "./sdef";
import { getGraphQLType, nullable } from "./types";
import camelCase from "camelcase";

export const collectMutationArgs = (c: {
  property?: PropertyDefinition[];
  element?: ElementDefinition[];
  contents?: ContentDefinition[];
}) => {
  const properties: InputValueDefinitionNode[] = (c.property ?? [])
    .filter((f) => !("access" in f.$ && f.$.access === "r"))
    .map((t) => {
      return {
        kind: Kind.INPUT_VALUE_DEFINITION,
        name: {
          kind: Kind.NAME,
          value: camelCase(t.$.name),
        },
        type: nullable(getGraphQLType(t)),
      };
    });

  return properties;
};
