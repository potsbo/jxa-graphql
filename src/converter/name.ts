import { Kind, NameNode } from "graphql";
import camelCase from "camelcase";

export const name = (n: string, opts: { pascalCase?: boolean } = {}): NameNode => {
  return {
    kind: Kind.NAME,
    value: camelCase(n, opts),
  };
};
