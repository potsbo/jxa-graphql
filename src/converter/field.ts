import {
  ConstDirectiveNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  Kind,
  StringValueNode,
  TypeNode,
} from "graphql";
import { ContentDefinition, ElementDefinition, PropertyDefinition } from "./sdef";
import { getGraphQLType, named, nonNull } from "./types";
import { name } from "./name";

const CONNECTION_TYPE_NAME = "Connection";

export const collectFieldsDefinitions = (c: {
  property?: PropertyDefinition[];
  element?: ElementDefinition[];
  contents?: ContentDefinition[];
}) => {
  const properties: FieldDefinitionNode[] = (c.property ?? []).map((t) => {
    if (t.$.code === "ID  ") {
      return field("id", nonNull("ID"), {
        description: t.$.description,
        directives: [
          {
            kind: Kind.DIRECTIVE,
            name: name("extractFromObjectDisplayName"),
          },
        ],
      });
    }
    return field(t.$.name, getGraphQLType(t), { description: t.$.description });
  });

  const elements: FieldDefinitionNode[] = (c.element ?? []).map((e) => {
    return field(`${e.$.type}s`, nonNull(named(e.$.type, CONNECTION_TYPE_NAME)), {
      description: e.$.description,
      arguments: [
        {
          kind: Kind.INPUT_VALUE_DEFINITION,
          name: name("whose"),
          type: named("Condition"),
        },
        {
          kind: Kind.INPUT_VALUE_DEFINITION,
          name: name("after"),
          type: named("ID"),
        },
        {
          kind: Kind.INPUT_VALUE_DEFINITION,
          name: name("first"),
          type: named("Int"),
        },
      ],
    });
  });

  const contents = (c.contents ?? []).map((ctnt): FieldDefinitionNode => {
    return field(ctnt.$.name, nonNull(ctnt.$.type), { description: ctnt.$.description });
  });
  // TODO: respond-to
  return properties.concat(elements).concat(contents);
};

export const field = (
  fieldName: string,
  type: TypeNode,
  opts?: {
    description?: string;
    arguments?: readonly InputValueDefinitionNode[];
    directives?: readonly ConstDirectiveNode[];
  }
): FieldDefinitionNode => {
  const desc: StringValueNode | undefined = opts?.description
    ? {
        kind: Kind.STRING,
        value: opts?.description,
        block: true,
      }
    : undefined;
  return {
    kind: Kind.FIELD_DEFINITION,
    name: name(fieldName),
    type,
    description: desc,
    arguments: opts?.arguments,
    directives: opts?.directives,
  };
};
