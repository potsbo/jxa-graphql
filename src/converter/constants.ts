import { InterfaceTypeDefinitionNode, Kind } from "graphql";
import { field } from "./field";
import { name } from "./name";
import { list, named, nonNull } from "./types";

export const CONNECTION_TYPE_NAME = "Connection";
export const EDGE_TYPE_NAME = "Edge";
export const NODE_TYPE_NAME = "Node";
export const INTERFACE_SUFFIX = "Interface";

export const NodeInterface: InterfaceTypeDefinitionNode = {
  kind: Kind.INTERFACE_TYPE_DEFINITION,
  name: name(NODE_TYPE_NAME, { pascalCase: true }),
  fields: [field("id", nonNull("ID"))],
};

// https://relay.dev/graphql/connections.htm#sec-Edge-Types
export const EdgeInterface: InterfaceTypeDefinitionNode = {
  kind: Kind.INTERFACE_TYPE_DEFINITION,
  name: name(EDGE_TYPE_NAME, { pascalCase: true }),
  fields: [field("node", nonNull("Node")), field("cursor", nonNull("String"))],
};

// https://relay.dev/graphql/connections.htm#sec-Connection-Types
export const ConnectionInterface: InterfaceTypeDefinitionNode = {
  kind: Kind.INTERFACE_TYPE_DEFINITION,
  name: name(CONNECTION_TYPE_NAME, { pascalCase: true }),
  fields: [
    field("edges", nonNull(list(nonNull(EDGE_TYPE_NAME)))),
    field("pageInfo", nonNull("PageInfo")),
    // https://developer.apple.com/library/archive/releasenotes/InterapplicationCommunication/RN-JavaScriptForAutomation/Articles/OSX10-10.html
    {
      ...field("byId", named("Node")),
      arguments: [
        {
          kind: Kind.INPUT_VALUE_DEFINITION,
          name: name("id"),
          type: nonNull("ID"),
        },
      ],
    },
  ],
};