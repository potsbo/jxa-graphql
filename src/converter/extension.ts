import { FieldDefinitionNode, Kind, ObjectTypeExtensionNode } from "graphql";
import camelCase from "camelcase";
import { collectFieldsDefinitions } from "./field";
import { ClassExtensionDefinition, Environment } from "./sdef";

export class ExtensionBuilder {
  private e: ClassExtensionDefinition;
  readonly fields: FieldDefinitionNode[];
  readonly extends: string;

  constructor(e: ClassExtensionDefinition) {
    this.e = e;
    this.fields = collectFieldsDefinitions(this.e);
    this.extends = e.$.extends;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  build = (_: Environment): ObjectTypeExtensionNode[] => {
    return [
      {
        kind: Kind.OBJECT_TYPE_EXTENSION,
        name: {
          kind: Kind.NAME,
          value: camelCase(this.e.$.extends, { pascalCase: true }),
        },
        fields: this.fields,
      },
    ];
  };
}
