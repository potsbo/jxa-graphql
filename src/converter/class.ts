import { DefinitionNode, FieldDefinitionNode, InterfaceTypeDefinitionNode, Kind } from "graphql";
import camelCase from "camelcase";
import { collectFieldsDefinitions, field } from "./field";
import { ClassDefinition, Environment } from "./sdef";
import { named as named, nonNull, list } from "./types";
import { EDGE_TYPE_NAME, CONNECTION_TYPE_NAME, NodeInterface } from "./constants";
import { name } from "./name";
import { objectType } from "./object";

const complementId = (fields: FieldDefinitionNode[]): FieldDefinitionNode[] => {
  if (fields.some((f) => f.name.value === "id")) {
    return fields;
  }

  return [
    ...fields,
    field("id", nonNull("ID"), {
      directives: [
        {
          kind: Kind.DIRECTIVE,
          name: name("extractFromObjectDisplayName"),
        },
      ],
    }),
  ];
};

export class ClassBuilder {
  private c: ClassDefinition;
  private fields: FieldDefinitionNode[];
  constructor(c: ClassDefinition) {
    this.c = c;
    this.fields = collectFieldsDefinitions(this.c);
  }
  private getBaseTypeName = () => camelCase(this.c.$.name, { pascalCase: true });
  private getClassName = () => this.c.$.name;
  private getInherits = () => this.c.$.inherits;
  private getInterfaceName = () => `${this.getBaseTypeName()}Interface`;
  private getInterfaced = (fields: FieldDefinitionNode[]): InterfaceTypeDefinitionNode => {
    return {
      kind: Kind.INTERFACE_TYPE_DEFINITION,
      fields,
      name: name(this.getInterfaceName(), { pascalCase: true }),
      interfaces: [named(NodeInterface.name.value)],
    };
  };
  private getAncestors = (classBuilders: ClassBuilder[]): ClassBuilder[] => {
    const inherits = this.getInherits();
    const parent = classBuilders.find((t) => t.getClassName() === inherits);
    if (inherits !== undefined && parent === undefined) {
      /* istanbul ignore next */
      throw new Error("parent not found");
    }
    if (parent === undefined) {
      return [];
    }
    return [parent, ...parent.getAncestors(classBuilders)];
  };
  build = ({ override, builders }: Environment): DefinitionNode[] => {
    const typeName = camelCase(this.c.$.name, { pascalCase: true });

    // TODO: if compatible override given, try to merge
    if (override?.definitions.some((d) => "name" in d && d.name?.value === typeName)) {
      return [];
    }

    const classBuilders = builders.filter((b): b is ClassBuilder => b instanceof ClassBuilder);
    const ancestors = this.getAncestors(classBuilders);

    const inherited = classBuilders.map((c) => c.getInherits()).some((c): c is string => c === this.getClassName());

    const fields = complementId(
      [this, ...ancestors].reduce((acum: FieldDefinitionNode[], cur) => {
        return [...acum, ...cur.fields];
      }, [])
    );

    const interfaces: string[] = [NodeInterface.name.value, ...ancestors.map((a) => a.getInterfaceName())];
    if (inherited) {
      interfaces.push(this.getInterfaceName());
    }

    const classDef = objectType(typeName, fields, { description: this.c.$.description, interfaces });
    const edgeDef = objectType(
      `${typeName}${EDGE_TYPE_NAME}`,
      [field("cursor", nonNull("String")), field("node", nonNull(inherited ? this.getInterfaceName() : typeName))],
      { interfaces: [EDGE_TYPE_NAME] }
    );
    const connectionDef = objectType(
      `${typeName}${CONNECTION_TYPE_NAME}`,
      [
        field("byId", named(inherited ? this.getInterfaceName() : typeName), {
          arguments: [
            {
              kind: Kind.INPUT_VALUE_DEFINITION,
              name: name("id"),
              type: nonNull("ID"),
            },
          ],
        }),
        field("edges", nonNull(list(nonNull(`${typeName}${EDGE_TYPE_NAME}`)))),
        field("pageInfo", nonNull("PageInfo")),
      ],
      {
        interfaces: [CONNECTION_TYPE_NAME],
      }
    );
    return [connectionDef, edgeDef, classDef, this.getInterfaced(fields)];
  };
}
