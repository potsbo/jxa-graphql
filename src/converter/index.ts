import execa from "execa";
import { parseStringPromise } from "xml2js";
import { print, lexicographicSortSchema, buildASTSchema, DocumentNode, DefinitionNode } from "graphql";
import { pruneSchema } from "@graphql-tools/utils";
import { Sdef } from "./sdef";
import { ConnectionInterface, EdgeInterface, NodeInterface } from "./constants";
import gql from "graphql-tag";
import { normalize } from "../normalize";
import { parseSuites } from "./suite";

const getBuilders = async (appPath: string) => {
  const sdefCmdResult = await execa("sdef", [appPath]);
  const sdef = (await parseStringPromise(sdefCmdResult.stdout)) as Sdef;
  return parseSuites(sdef);
};

const getBuildersFromFile = async (filepath: string) => {
  const content = await execa("cat", [filepath]);
  const sdef = (await parseStringPromise(content.stdout)) as Sdef;
  return parseSuites(sdef);
};

export const build = async (appPath: string, override?: DocumentNode) => {
  let { builders, includes } = await getBuilders(appPath);
  while (includes.length > 0) {
    for (const i of includes) {
      const { builders: bs, includes: is } = await getBuildersFromFile(i);
      includes = is;
      builders = [...builders, ...bs];
    }
  }

  const env = { builders, override };

  const definitions: DefinitionNode[] = [ConnectionInterface, EdgeInterface, NodeInterface];
  builders.forEach((b) => {
    definitions.push(...b.build(env));
  });

  const schema = gql`
    type Query {
      application: Application!
    }
    type Mutation

    ${definitions.map(print).join("\n")}

    # https://relay.dev/graphql/connections.htm#sec-undefined.PageInfo
    type PageInfo {
      hasPreviousPage: Boolean!
      hasNextPage: Boolean!
      startCursor: String!
      endCursor: String!
    }

    directive @recordType on OBJECT
    directive @extractFromObjectDisplayName on FIELD_DEFINITION

    input Condition {
      enabled: Boolean! = true
      field: String
      operands: [Condition!]
      operator: String! = "="
      value: String! = "true"
    }

    ${override ? print(override) : ""}
  `;

  return lexicographicSortSchema(pruneSchema(buildASTSchema(normalize(schema))));
};
