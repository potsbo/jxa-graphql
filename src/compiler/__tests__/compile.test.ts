import { buildExecutionContext, ExecutionContext } from "graphql/execution/execute";
import prettier from "prettier";
import { GraphQLError } from "graphql";
import gql from "graphql-tag";
import { compile } from "../index";
import { build } from "../../converter";

const validateExecontext = (obj: ExecutionContext | readonly GraphQLError[]): obj is ExecutionContext => {
  return "operation" in obj;
};

const schema = build("/Applications/OmniFocus.app");

const TaskViewModelFragmentDoc = gql`
  fragment TaskViewModel on TaskInterface {
    name
    id
    effectiveDueDate
    completed
    effectivelyCompleted
    containingProject {
      id
      name
    }
    flagged
  }
`;

test("query for Connection", async () => {
  const document = gql`
    query Hoge {
      application {
        defaultDocument {
          projects {
            pageInfo {
              hasNextPage
            }
            edges {
              cursor
              node {
                name
              }
            }
          }
        }
      }
    }
  `;
  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
  });
  if (!validateExecontext(exeContext)) {
    fail();
  }

  expect(prettier.format(compile("OmniFocus", exeContext), { parser: "babel" })).toMatchSnapshot();
});

test("query with inline fragment", async () => {
  const document = gql`
    query {
      application {
        defaultDocument {
          folders {
            edges {
              node {
                name
                sections {
                  edges {
                    node {
                      ... on Project {
                        completed
                      }
                      ... on Folder {
                        name
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  `;
  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
  });
  if (!validateExecontext(exeContext)) {
    fail();
  }

  expect(prettier.format(compile("OmniFocus", exeContext), { parser: "babel" })).toMatchSnapshot();
});

test("query with project interface", async () => {
  const document = gql`
    query {
      application {
        defaultDocument {
          projects {
            byId(id: "hZoaSakOnG4") {
              rootTask {
                id
              }
            }
          }
        }
      }
    }
  `;
  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
  });
  if (!validateExecontext(exeContext)) {
    fail();
  }

  expect(prettier.format(compile("OmniFocus", exeContext), { parser: "babel" })).toMatchSnapshot();
});

test("query with pageInfo", async () => {
  const document = gql`
    query {
      application {
        defaultDocument {
          projects(first: 10, after: "SOME-ID") {
            pageInfo {
              hasPreviousPage
              hasNextPage
              startCursor
              endCursor
            }
          }
        }
      }
    }
  `;
  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
  });
  if (!validateExecontext(exeContext)) {
    fail();
  }

  expect(prettier.format(compile("OmniFocus", exeContext), { parser: "babel" })).toMatchSnapshot();
});

test("query with fragment", async () => {
  const document = gql`
    query GetTasksInProject($projectId: ID!) {
      application {
        defaultDocument {
          projects {
            byId(id: $projectId) {
              rootTask {
                tasks {
                  edges {
                    node {
                      ...TaskViewModel
                      repetition {
                        unit
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
    ${TaskViewModelFragmentDoc}
  `;

  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
    variableValues: { projectId: "foobar" },
  });
  if (!validateExecontext(exeContext)) {
    console.log(exeContext);
  }

  expect(prettier.format(compile("OmniFocus", exeContext as any), { parser: "babel" })).toMatchSnapshot();
});

test("typename", async () => {
  const document = gql`
    query GetTasksInProject {
      application {
        perspectiveNames
        defaultDocument {
          sections {
            edges {
              node {
                ... on Project {
                  name
                  effectiveStatus
                }
                ... on Folder {
                  name
                }
              }
            }
          }
        }
      }
    }
  `;

  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
  });
  if (!validateExecontext(exeContext)) {
    console.log(exeContext);
  }

  expect(prettier.format(compile("OmniFocus", exeContext as any), { parser: "babel" })).toMatchSnapshot();
});

test("explicit typename", async () => {
  const document = gql`
    query {
      application {
        defaultDocument {
          sections {
            edges {
              node {
                id
                __typename
              }
            }
          }
        }
      }
    }
  `;

  const exeContext = buildExecutionContext({
    schema: await schema,
    document: document,
  });
  if (!validateExecontext(exeContext)) {
    console.log(exeContext);
  }

  expect(prettier.format(compile("OmniFocus", exeContext as any), { parser: "babel" })).toMatchSnapshot();
});
