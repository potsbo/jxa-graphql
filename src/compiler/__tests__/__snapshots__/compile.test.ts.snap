// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query for Connection 1`] = `
"function extractClass(obj) {
  const s = ObjectSpecifier.classOf(eval(Automation.getDisplayString(obj)));
  return (s.match(/[a-zA-Z0-9]+/g) || [])
    .map((w) => \`\${w[0].toUpperCase()}\${w.slice(1)}\`)
    .join(\\"\\");
}
function extractId(obj) {
  const spec = Automation.getDisplayString(obj);
  if (spec === undefined) {
    return null;
  }
  let inParen = false;
  let inStr = false;
  let ret = \\"\\";
  let nextEscape = null;
  for (let i = 0; i < spec.length; i++) {
    const c = spec[i];
    const inEscape = nextEscape === i;
    if (c === \\"\\\\\\\\\\") {
      nextEscape = i + 1;
      continue;
    }
    if (!inEscape && c === '\\"') {
      inStr = !inStr;
      continue;
    }
    if (!inStr) {
      if (c === \\"(\\") {
        inParen = true;
        ret = \\"\\";
        continue;
      }
      if (c === \\")\\") {
        inParen = false;
        continue;
      }
    }
    ret = \`\${ret}\${c}\`;
  }
  return ret !== \\"\\" ? ret : null;
}
function paginate(nodes, { first, after }, getId) {
  const afterIndex =
    after === undefined
      ? undefined
      : nodes.findIndex((n) => getId(n) === after);
  if (afterIndex === -1) {
    return [];
  }
  const start = afterIndex === undefined ? 0 : afterIndex + 1;
  const end = first !== undefined ? start + first : undefined;
  return nodes.slice(start, end);
}
const app = Application(\\"OmniFocus\\");
JSON.stringify({
  result: {
    defaultDocument: {
      projects: (() => {
        const allNodes = app.defaultDocument().projects();
        const nodes = allNodes;
        return {
          edges: nodes.map((elm) => {
            return {
              cursor: extractId(elm),
              node: { __typename: extractClass(elm), name: elm.name() },
            };
          }),
          pageInfo: {
            hasNextPage:
              extractId(nodes[nodes.length - 1]) !==
              extractId(allNodes[allNodes.length - 1]),
          },
        };
      })(),
    },
  },
});
"
`;

exports[`query with fragment 1`] = `
"function extractClass(obj) {
  const s = ObjectSpecifier.classOf(eval(Automation.getDisplayString(obj)));
  return (s.match(/[a-zA-Z0-9]+/g) || [])
    .map((w) => \`\${w[0].toUpperCase()}\${w.slice(1)}\`)
    .join(\\"\\");
}
function extractId(obj) {
  const spec = Automation.getDisplayString(obj);
  if (spec === undefined) {
    return null;
  }
  let inParen = false;
  let inStr = false;
  let ret = \\"\\";
  let nextEscape = null;
  for (let i = 0; i < spec.length; i++) {
    const c = spec[i];
    const inEscape = nextEscape === i;
    if (c === \\"\\\\\\\\\\") {
      nextEscape = i + 1;
      continue;
    }
    if (!inEscape && c === '\\"') {
      inStr = !inStr;
      continue;
    }
    if (!inStr) {
      if (c === \\"(\\") {
        inParen = true;
        ret = \\"\\";
        continue;
      }
      if (c === \\")\\") {
        inParen = false;
        continue;
      }
    }
    ret = \`\${ret}\${c}\`;
  }
  return ret !== \\"\\" ? ret : null;
}
function paginate(nodes, { first, after }, getId) {
  const afterIndex =
    after === undefined
      ? undefined
      : nodes.findIndex((n) => getId(n) === after);
  if (afterIndex === -1) {
    return [];
  }
  const start = afterIndex === undefined ? 0 : afterIndex + 1;
  const end = first !== undefined ? start + first : undefined;
  return nodes.slice(start, end);
}
const projectId = \\"foobar\\";
const app = Application(\\"OmniFocus\\");
JSON.stringify({
  result: {
    defaultDocument: {
      projects: (() => {
        const allNodes = app.defaultDocument().projects();
        const nodes = allNodes;
        return {
          byId: app.defaultDocument().projects.byId(projectId)
            ? {
                __typename: extractClass(
                  app.defaultDocument().projects.byId(projectId)
                ),
                rootTask: {
                  tasks: (() => {
                    const allNodes = app
                      .defaultDocument()
                      .projects.byId(projectId)
                      .rootTask()
                      .tasks();
                    const nodes = allNodes;
                    return {
                      edges: nodes.map((elm) => {
                        return {
                          node: {
                            __typename: extractClass(elm),
                            completed: elm.completed(),
                            containingProject: elm.containingProject()
                              ? {
                                  id: extractId(elm.containingProject()),
                                  name: elm.containingProject().name(),
                                }
                              : undefined,
                            effectiveDueDate: elm.effectiveDueDate(),
                            effectivelyCompleted: elm.effectivelyCompleted(),
                            flagged: elm.flagged(),
                            id: extractId(elm),
                            name: elm.name(),
                            repetition: elm.repetition()
                              ? {
                                  unit: elm
                                    .repetition()
                                    .unit()
                                    ?.toUpperCase()
                                    .replaceAll(\\" \\", \\"_\\"),
                                }
                              : undefined,
                          },
                        };
                      }),
                    };
                  })(),
                },
              }
            : undefined,
        };
      })(),
    },
  },
});
"
`;

exports[`query with inline fragment 1`] = `
"function extractClass(obj) {
  const s = ObjectSpecifier.classOf(eval(Automation.getDisplayString(obj)));
  return (s.match(/[a-zA-Z0-9]+/g) || [])
    .map((w) => \`\${w[0].toUpperCase()}\${w.slice(1)}\`)
    .join(\\"\\");
}
function extractId(obj) {
  const spec = Automation.getDisplayString(obj);
  if (spec === undefined) {
    return null;
  }
  let inParen = false;
  let inStr = false;
  let ret = \\"\\";
  let nextEscape = null;
  for (let i = 0; i < spec.length; i++) {
    const c = spec[i];
    const inEscape = nextEscape === i;
    if (c === \\"\\\\\\\\\\") {
      nextEscape = i + 1;
      continue;
    }
    if (!inEscape && c === '\\"') {
      inStr = !inStr;
      continue;
    }
    if (!inStr) {
      if (c === \\"(\\") {
        inParen = true;
        ret = \\"\\";
        continue;
      }
      if (c === \\")\\") {
        inParen = false;
        continue;
      }
    }
    ret = \`\${ret}\${c}\`;
  }
  return ret !== \\"\\" ? ret : null;
}
function paginate(nodes, { first, after }, getId) {
  const afterIndex =
    after === undefined
      ? undefined
      : nodes.findIndex((n) => getId(n) === after);
  if (afterIndex === -1) {
    return [];
  }
  const start = afterIndex === undefined ? 0 : afterIndex + 1;
  const end = first !== undefined ? start + first : undefined;
  return nodes.slice(start, end);
}
const app = Application(\\"OmniFocus\\");
JSON.stringify({
  result: {
    defaultDocument: {
      folders: (() => {
        const allNodes = app.defaultDocument().folders();
        const nodes = allNodes;
        return {
          edges: nodes.map((elm) => {
            return {
              node: {
                __typename: extractClass(elm),
                name: elm.name(),
                sections: (() => {
                  const allNodes = elm.sections();
                  const nodes = allNodes;
                  return {
                    edges: nodes.map((elm) => {
                      return {
                        node: {
                          __typename: extractClass(elm),
                          ...(() => {
                            return extractClass(elm) === \\"Folder\\"
                              ? {
                                  name: elm.name(),
                                  __typename: \\"Folder\\",
                                }
                              : {};
                          })(),
                          ...(() => {
                            return extractClass(elm) === \\"Project\\"
                              ? {
                                  completed: elm.completed(),
                                  __typename: \\"Project\\",
                                }
                              : {};
                          })(),
                        },
                      };
                    }),
                  };
                })(),
              },
            };
          }),
        };
      })(),
    },
  },
});
"
`;

exports[`query with pageInfo 1`] = `
"function extractClass(obj) {
  const s = ObjectSpecifier.classOf(eval(Automation.getDisplayString(obj)));
  return (s.match(/[a-zA-Z0-9]+/g) || [])
    .map((w) => \`\${w[0].toUpperCase()}\${w.slice(1)}\`)
    .join(\\"\\");
}
function extractId(obj) {
  const spec = Automation.getDisplayString(obj);
  if (spec === undefined) {
    return null;
  }
  let inParen = false;
  let inStr = false;
  let ret = \\"\\";
  let nextEscape = null;
  for (let i = 0; i < spec.length; i++) {
    const c = spec[i];
    const inEscape = nextEscape === i;
    if (c === \\"\\\\\\\\\\") {
      nextEscape = i + 1;
      continue;
    }
    if (!inEscape && c === '\\"') {
      inStr = !inStr;
      continue;
    }
    if (!inStr) {
      if (c === \\"(\\") {
        inParen = true;
        ret = \\"\\";
        continue;
      }
      if (c === \\")\\") {
        inParen = false;
        continue;
      }
    }
    ret = \`\${ret}\${c}\`;
  }
  return ret !== \\"\\" ? ret : null;
}
function paginate(nodes, { first, after }, getId) {
  const afterIndex =
    after === undefined
      ? undefined
      : nodes.findIndex((n) => getId(n) === after);
  if (afterIndex === -1) {
    return [];
  }
  const start = afterIndex === undefined ? 0 : afterIndex + 1;
  const end = first !== undefined ? start + first : undefined;
  return nodes.slice(start, end);
}
const app = Application(\\"OmniFocus\\");
JSON.stringify({
  result: {
    defaultDocument: {
      projects: (() => {
        const allNodes = app.defaultDocument().projects();
        const nodes = paginate(
          allNodes,
          { first: 10, after: \\"SOME-ID\\" },
          extractId
        );
        return {
          pageInfo: {
            endCursor: extractId(nodes[nodes.length - 1]),
            hasNextPage:
              extractId(nodes[nodes.length - 1]) !==
              extractId(allNodes[allNodes.length - 1]),
            hasPreviousPage: extractId(nodes[0]) !== extractId(allNodes[0]),
            startCursor: extractId(nodes[0]),
          },
        };
      })(),
    },
  },
});
"
`;

exports[`query with project interface 1`] = `
"function extractClass(obj) {
  const s = ObjectSpecifier.classOf(eval(Automation.getDisplayString(obj)));
  return (s.match(/[a-zA-Z0-9]+/g) || [])
    .map((w) => \`\${w[0].toUpperCase()}\${w.slice(1)}\`)
    .join(\\"\\");
}
function extractId(obj) {
  const spec = Automation.getDisplayString(obj);
  if (spec === undefined) {
    return null;
  }
  let inParen = false;
  let inStr = false;
  let ret = \\"\\";
  let nextEscape = null;
  for (let i = 0; i < spec.length; i++) {
    const c = spec[i];
    const inEscape = nextEscape === i;
    if (c === \\"\\\\\\\\\\") {
      nextEscape = i + 1;
      continue;
    }
    if (!inEscape && c === '\\"') {
      inStr = !inStr;
      continue;
    }
    if (!inStr) {
      if (c === \\"(\\") {
        inParen = true;
        ret = \\"\\";
        continue;
      }
      if (c === \\")\\") {
        inParen = false;
        continue;
      }
    }
    ret = \`\${ret}\${c}\`;
  }
  return ret !== \\"\\" ? ret : null;
}
function paginate(nodes, { first, after }, getId) {
  const afterIndex =
    after === undefined
      ? undefined
      : nodes.findIndex((n) => getId(n) === after);
  if (afterIndex === -1) {
    return [];
  }
  const start = afterIndex === undefined ? 0 : afterIndex + 1;
  const end = first !== undefined ? start + first : undefined;
  return nodes.slice(start, end);
}
const app = Application(\\"OmniFocus\\");
JSON.stringify({
  result: {
    defaultDocument: {
      projects: (() => {
        const allNodes = app.defaultDocument().projects();
        const nodes = allNodes;
        return {
          byId: app.defaultDocument().projects.byId(\\"hZoaSakOnG4\\")
            ? {
                __typename: extractClass(
                  app.defaultDocument().projects.byId(\\"hZoaSakOnG4\\")
                ),
                rootTask: {
                  id: extractId(
                    app
                      .defaultDocument()
                      .projects.byId(\\"hZoaSakOnG4\\")
                      .rootTask()
                  ),
                },
              }
            : undefined,
        };
      })(),
    },
  },
});
"
`;

exports[`typename 1`] = `
"function extractClass(obj) {
  const s = ObjectSpecifier.classOf(eval(Automation.getDisplayString(obj)));
  return (s.match(/[a-zA-Z0-9]+/g) || [])
    .map((w) => \`\${w[0].toUpperCase()}\${w.slice(1)}\`)
    .join(\\"\\");
}
function extractId(obj) {
  const spec = Automation.getDisplayString(obj);
  if (spec === undefined) {
    return null;
  }
  let inParen = false;
  let inStr = false;
  let ret = \\"\\";
  let nextEscape = null;
  for (let i = 0; i < spec.length; i++) {
    const c = spec[i];
    const inEscape = nextEscape === i;
    if (c === \\"\\\\\\\\\\") {
      nextEscape = i + 1;
      continue;
    }
    if (!inEscape && c === '\\"') {
      inStr = !inStr;
      continue;
    }
    if (!inStr) {
      if (c === \\"(\\") {
        inParen = true;
        ret = \\"\\";
        continue;
      }
      if (c === \\")\\") {
        inParen = false;
        continue;
      }
    }
    ret = \`\${ret}\${c}\`;
  }
  return ret !== \\"\\" ? ret : null;
}
function paginate(nodes, { first, after }, getId) {
  const afterIndex =
    after === undefined
      ? undefined
      : nodes.findIndex((n) => getId(n) === after);
  if (afterIndex === -1) {
    return [];
  }
  const start = afterIndex === undefined ? 0 : afterIndex + 1;
  const end = first !== undefined ? start + first : undefined;
  return nodes.slice(start, end);
}
const app = Application(\\"OmniFocus\\");
JSON.stringify({
  result: {
    defaultDocument: {
      sections: (() => {
        const allNodes = app.defaultDocument().sections();
        const nodes = allNodes;
        return {
          edges: nodes.map((elm) => {
            return {
              node: {
                __typename: extractClass(elm),
                ...(() => {
                  return extractClass(elm) === \\"Folder\\"
                    ? {
                        name: elm.name(),
                        __typename: \\"Folder\\",
                      }
                    : {};
                })(),
                ...(() => {
                  return extractClass(elm) === \\"Project\\"
                    ? {
                        effectiveStatus: elm
                          .effectiveStatus()
                          ?.toUpperCase()
                          .replaceAll(\\" \\", \\"_\\"),
                        name: elm.name(),
                        __typename: \\"Project\\",
                      }
                    : {};
                })(),
              },
            };
          }),
        };
      })(),
    },
    perspectiveNames: app.perspectiveNames(),
  },
});
"
`;
